### PR TITLE
Add bom_sniffing option to DecodeReaderBytesBuilder

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
https://github.com/BurntSushi/ripgrep/issues/1207

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -161,8 +163,9 @@ impl DecodeReaderBytesBuilder {
         }
         let encoding = self.encoding
             .map(|enc| enc.new_decoder_with_bom_removal());
-        // No need to do BOM detection if we have an explicit encoding.
-        let has_detected = !self.bom_override && encoding.is_some();
+
+        // No need to do BOM detection if we opt out of it or have an explicit encoding.
+        let has_detected = !self.bom_sniffing || !self.bom_override && encoding.is_some();
```
Comment (BurntSushi) [inline src/lib.rs]: Can you add parens here to indicate precedent?

File: src/lib.rs [hunk 2]
```diff
@@ -312,6 +315,23 @@ impl DecodeReaderBytesBuilder {
...
         self
     }
+
+    /// Enable BOM sniffing
+    ///
+    /// When this is enabled and an explicit encoding is not set, the decoder will try to detect
+    /// the encoding with BOM.
+    ///
+    /// When this is disabled and an explicit encoding is not set, the decoder will treat the input
+    /// as raw bytes.
```
Comment (BurntSushi) [inline src/lib.rs]: I would add, "The bytes will be passed through unchanged, including any BOM that may be present."

File: src/lib.rs (hunk 1 above)
Comment (LesnyRumcajs) [reply src/lib.rs]: Done!

File: src/lib.rs (hunk 2 above)
Comment (LesnyRumcajs) [reply src/lib.rs]: Done!

File: src/lib.rs [hunk 3]
```diff
@@ -161,8 +163,9 @@ impl DecodeReaderBytesBuilder {
         }
         let encoding = self.encoding
             .map(|enc| enc.new_decoder_with_bom_removal());
-        // No need to do BOM detection if we have an explicit encoding.
-        let has_detected = !self.bom_override && encoding.is_some();
+
+        // No need to do BOM detection if we opt out of it or have an explicit encoding.
+        let has_detected = !self.bom_sniffing || (!self.bom_override && encoding.is_some());
```
Comment (BurntSushi) [inline src/lib.rs]: Sorry, I just noticed this: can you make sure all new code introduced in this is wrapped to 79 columns inclusive? It should follow the existing style of the code. This line in particular is longer than 79.

Comment (LesnyRumcajs) [reply src/lib.rs]: Done. I wanted to use `cargo fmt` at first but didn't know if this repository follows it's rules. Perhaps this tool can be used in the future?

Comment (BurntSushi) [reply src/lib.rs]: It'd be nice, but no, I don't think rustfmt is in a place yet for me to adopt it. It does too many sub-optimal things with the formatting.

**Comment by BurntSushi** [comment]
This PR is now in `encoding_rs_io 0.1.5` on crates.io. I've also bumped the dependency in ripgrep for you: https://github.com/BurntSushi/ripgrep/commit/09139721047b1cda6ad88dbf89dc5fa74c66a3a2

**Comment by LesnyRumcajs** [comment]
@BurntSushi  Great, thanks. I will work on the "plumbing" part now. :)

**Pull request merged**